lib/cn/csched_sp3.c

```diff
@@ -932,37 +931,45 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
             sp3_node_remove(sp, spn, wtype_scatter);
         }
 
-        /* Leaf nodes sorted by pct capacity and secondarily by
-         * number of keys.  If the node's size doesn't exceed the "split_pct"
-         * threshold then we check to see if the number of keys exceeds the
-         * "split_keys" threshold.  If so, we insert this node into the tree
-         * with "split_pct" capacity to ensure it gets split or compacted.
+        /* Leaf nodes sorted by pct capacity and secondarily by number of keys.
+         * If the node's size doesn't exceed the "split_pct" threshold then
+         * we check to see if the number of keys exceeds the "split_keys"
+         * threshold.  If so, we insert this node into the tree using the
+         * "split_pct" capacity to ensure it gets split or compacted.
+         *
+         * Node splits are rare, but when a node is ready to split it should
+         * be done as soon as possible.  Hence, splits prevent other "big"
+         * compaction jobs from starting which could otherwise indefinitely
+         * starve a split.
          */
         if (tn->tn_ns.ns_pcap >= sp->thresh.lcomp_split_pct && jobs < 1) {
             const uint64_t weight = ((uint64_t)tn->tn_ns.ns_pcap << 32) | keys;
+            uint64_t split_keys = (uint64_t)sp->thresh.lcomp_split_keys << 22;
```

Comment:
Where do these shift values come from?

Reply:
See the comments for the variables and related defines.  Actually, the threshold structure is unnecessarily compacted which causes stuff like this and casting to larger type.  If we uncompacted it we could eliminate all this silly mess.  I'll open a ticket..

Reply:
https://jira.micron.com/jira/browse/NFSE-5332
This has been irritating me for quite a while, thanks for pointing it out!